### PR TITLE
Add multi-step add job function

### PR DESCRIPTION
**Pull request by jraede**

### CODE REVIEW

File: lib/dispatch.js [hunk 1]
```diff
@@ -483,4 +483,106 @@ export default class Dispatch {
...
+    };
+
+    return new Promise((resolve, reject) => {
+      createCustomer()
+        .then(() => {
+          return createJob();
+        })
+        .then(() => {
+          return createAppointment();
+        })
```
Comment (alvincrespo) [inline lib/dispatch.js]: Couldn't you inline the above functions?

```
.then(() => createJob())
.then(() => createAppointment())
```

File: lib/dispatch.js [hunk 2]
```diff
@@ -483,4 +483,106 @@ export default class Dispatch {
...
+   * error type has a `data` property, which in this case will be an object with `job`, `customer`,
+   * and `appointment` properties, denoting the error that happened in each step, if applicable.
+   *
+   * IF it succeeds, the returned promise will be RESOLVED with an object containing properties for
+   * `job`, `customer`, and `appointment`, which will be the returned values from the API as a
+   * result of creating each respective entity.
+   *
+   * @param  {Object} jobInfo         Job data
+   * @param  {Object} customerInfo    Customer data
+   * @param  {[Object]} appointmentInfo Appointment data
```
Comment (alvincrespo) [inline lib/dispatch.js]: Does this mean optional? `{[Object]}` - that looks like the param is an array of objects?

File: lib/dispatch.js [hunk 3]
```diff
@@ -483,4 +483,106 @@ export default class Dispatch {
...
+
+    const createAppointment = () => {
+      if (!appointmentInfo) {
+        return Promise.resolve();
+      }
+      return new Promise((resolve, reject) => {
+        this.entities.appointments.create(appointmentInfo)
+        .then(appointment => {
+          returnData.appointment = appointment;
+          resolve();
```
Comment (alvincrespo) [inline lib/dispatch.js]: Why do we not resolve with the returnData? That way you could then pass that down the promise chain.

File: lib/errors.js [hunk 4]
```diff
@@ -55,3 +55,14 @@ export class APIError {
...
+export class MultiActionError {
+  constructor(message, data) {
+    this.name = 'MultiActionError';
+    this.message = message;
+    this.data = data;
+    this.stack = (new Error()).stack;
+  }
+}
+
+makeError(MultiActionError);
```
Comment (alvincrespo) [inline lib/errors.js]: 👍

File: lib/dispatch.js [hunk 5]
```diff
@@ -483,4 +483,106 @@ export default class Dispatch {
...
+        .then(customer => {
+          returnData.customer = customer;
+          jobInfo.customer_id = customer.id;
+          resolve();
+        }).catch(err => {
+          errorData.customer = err;
+          reject(err);
+        });
+      });
+    };
```
Comment (alvincrespo) [inline lib/dispatch.js]: Couldn't these be placed outside of createAndScheduleJob as individual methods that could be, theoretically, exported and used elsewhere?

Comment (kswin) [inline lib/dispatch.js]: These methods are reusable by calling `this.entities.customers.create` and `this.entities.jobs.create(jobInfo)`. The promise success handler is specific to this function, so I think it makes sense to keep it as is.

File: lib/dispatch.js (hunk 3 above)
Comment (kswin) [inline lib/dispatch.js]: This is a case where a promise is used to denote that an action has completed vs. forwarding data - hence, no need to resolve with `returnData`

**Pull request merged**